edn/src/query.rs

```diff
@@ -0,0 +1,1253 @@
+    ///
+    /// > Datomic will attempt to push the or clause down until all necessary variables are bound,
+    /// > and will throw an exception if that is not possible.
+    ///
+    /// but that would render top-level `or` expressions (as used in Datomic's own examples!)
+    /// impossible, so we assume that this is an error in the documentation.
+    ///
+    /// All contained 'arms' in an `or` with implicit variables must bind the same vars.
+    Implicit,
+
+    /// `Explicit` means the variables in an `or-join` or `not-join` are explicitly listed,
+    /// specified with `required-vars` syntax.
+    ///
+    /// DataScript parses these as free, but allows (incorrectly) the use of more complicated
+    /// `rule-vars` syntax.
+    ///
+    /// Only the named variables will be unified with the enclosing query.
+    ///
+    /// Every 'arm' in an `or-join` must mention the entire set of explicit vars.
+    Explicit(BTreeSet<Variable>),
+}
+
+impl WhereClause {
+    pub fn is_pattern(&self) -> bool {
+        match self {
+            &WhereClause::Pattern(_) => true,
+            _ => false,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum OrWhereClause {
+    Clause(WhereClause),
+    And(Vec<WhereClause>),
+}
+
+impl OrWhereClause {
+    pub fn is_pattern_or_patterns(&self) -> bool {
+        match self {
+            &OrWhereClause::Clause(WhereClause::Pattern(_)) => true,
+            &OrWhereClause::And(ref clauses) => clauses.iter().all(|clause| clause.is_pattern()),
+            _ => false,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct OrJoin {
+    pub unify_vars: UnifyVars,
+    pub clauses: Vec<OrWhereClause>,
+
+    /// Caches the result of `collect_mentioned_variables`.
+    mentioned_vars: Option<BTreeSet<Variable>>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct NotJoin {
+    pub unify_vars: UnifyVars,
+    pub clauses: Vec<WhereClause>,
+}
+
+impl NotJoin {
+    pub fn new(unify_vars: UnifyVars, clauses: Vec<WhereClause>) -> NotJoin {
+        NotJoin {
+            unify_vars: unify_vars,
+            clauses: clauses,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct TypeAnnotation {
+    pub value_type: Keyword,
+    pub variable: Variable,
+}
+
+#[allow(dead_code)]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum WhereClause {
+    NotJoin(NotJoin),
+    OrJoin(OrJoin),
+    Pred(Predicate),
+    WhereFn(WhereFn),
+    RuleExpr,
+    Pattern(Pattern),
+    TypeAnnotation(TypeAnnotation),
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Eq, PartialEq)]
+pub struct ParsedFindQuery {
+    pub find_spec: FindSpec,
+    pub default_source: SrcVar,
+    pub with: Vec<Variable>,
+    pub in_vars: Vec<Variable>,
+    pub in_sources: BTreeSet<SrcVar>,
+    pub limit: Limit,
+    pub where_clauses: Vec<WhereClause>,
+    pub order: Option<Vec<Order>>,
+    // TODO: in_rules;
```

Comment:
Issue number?

Reply:
Historical, pretty sure it's untracked.  I think I'll remove it, since I very much doubt we'll ever implement Datalog rules.